### PR TITLE
Implemented SHA1 scalar function with the help of MessageDigest.

### PR DESCRIPTION
**Pull request by mrredlich**
Standard Tests added.

### CODE REVIEW

**Comment by chaudum** [comment]
Hi @mrredlich 
Thanks for your contribution! I've seen that you also have signed the CLA already.
To be able to merge your PR you'd also need to add an entry to the changelog (under the "unreleased" section".

Someone of our core developer ( @matriv or @seut ) will also have a look.

**Comment by mrredlich** [comment]
Hi @chaudum, @seut, @matriv.
Thanks for looking over it. It seems that one of the new JUnit tests failed in travis which is a little strange since they run perfectly fine on my machine (I just double checked again). Since it fails at the umlaut assertion I'm assuming that it's caused by the compiler's or system's language settings or something similar. I'm currently on holidays but will have a more thorough look at the problem when I'm back.

File: sql/src/main/java/io/crate/operation/scalar/string/Sha1Function.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import io.crate.operation.Input;
+import io.crate.operation.scalar.ScalarFunctionModule;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+public class Sha1Function extends Scalar<BytesRef, BytesRef>{
+	
+	public final static String NAME = "sha1";
+	private final FunctionInfo info;
+	private MessageDigest messageDigest = null;
```
Comment (matriv) [inline sql/src/main/java/io/crate/operation/scalar/string/Sha1Function.java]: Don't use messageDigest variable as a class field. It's not thread safe.
Instead create a new instance every time in evaluate().
See: http://stackoverflow.com/questions/17554998/need-thread-safe-messagedigest-in-java

File: sql/src/main/java/io/crate/operation/scalar/string/Sha1Function.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        if (stringValue == null) {
+            return null;
+        }
+                
+        //sha function executed to arg0 using apache commons
+        if( messageDigest == null ) {	
+	        try {
+	        	messageDigest = MessageDigest.getInstance("SHA");
+			} catch (NoSuchAlgorithmException e) {
+				e.printStackTrace();
```
Comment (matriv) [inline sql/src/main/java/io/crate/operation/scalar/string/Sha1Function.java]: Use ESLogger:
private final static ESLogger LOGGER = Loggers.getLogger(SQLOperations.class);
and
LOGGER.error("Message", e);

File: sql/src/test/java/io/crate/operation/scalar/string/Sha1DigestTest.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+import static io.crate.testing.TestingHelpers.isLiteral;
+
+import org.junit.Test;
+
+import io.crate.analyze.symbol.Literal;
+import io.crate.operation.scalar.AbstractScalarFunctionsTest;
+import io.crate.types.DataTypes;
+
+public class Sha1DigestTest extends AbstractScalarFunctionsTest{
```
Comment (matriv) [inline sql/src/test/java/io/crate/operation/scalar/string/Sha1DigestTest.java]: Use the same method as the class you're testing: Sha1FunctionTest.
Add space before {

**Comment by matriv** [comment]
Hi @mrredlich!
Please check my comments. Maybe the thread safety thing is responsible for the travis build failures.

**Comment by mrredlich** [comment]
Thanks Marios for the feedback. I've carried out a few (simple) performance tests to figure out what implementation would be the quickest. The results and routine is in the uncommented bit in Sha1FunctionTest.java
The test is still failing because some of the test variables in the are based on UTF-16 characters, e.g. £, ä, ö, ü, and Ö. Should I simply change these tests so only UTF-8 characters are tested?

**Comment by mikethebeer** [comment]
@mrredlich @matriv wouldn't it be better to use `DigestUtils` from Apache Commons for the SHA1 generation? https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/digest/DigestUtils.html

**Comment by mrredlich** [comment]
Hi @matriv @mikethebeer, 
I've done exactly that: I tested the different implementation (MessageDigest and DigestUtils) for their performance. The performance results can be found in Sha1FunctionTest.java as comment. The MessageDigest implementation seems to have been slightly quicker for bigger texts (see results) which is why I chose to use it. Btw. the other two implementations are still in the Sha1Function.java but as comment. It would be great to hear advice whether to remove this code (test results + unused implementations) or leave it in.  Also any advice on the UTF problem. Is crate simply working in an UTF-8 environment, i.e. not supporting UTF-16?

**Comment by seut** [comment]
Closing because planned to be implemented in a different way.

**Pull request closed without merging**